Sources/Swift/Tools/UIRedactBuilder.swift

```diff
@@ -141,46 +133,65 @@ class UIRedactBuilder {
         return image.imageAsset?.value(forKey: "_containingBundle") == nil
     }
     
-    private func mapRedactRegion(fromView view: UIView, to: CALayer, redacting: inout [RedactRegion], area: CGRect, redactText: Bool, redactImage: Bool) {
-        let rectInWindow = (view.layer.presentation() ?? view.layer).convert(view.bounds, to: to)
-        guard (redactImage || redactText) && area.intersects(rectInWindow) && !view.isHidden && view.alpha != 0 else { return }
+    private func mapRedactRegion(fromView view: UIView, redacting: inout [RedactRegion], rootFrame: CGRect, redactOptions: SentryRedactOptions, transform: CGAffineTransform) {
+        guard (redactOptions.redactAllImages || redactOptions.redactAllText) && !view.isHidden && view.alpha != 0 else { return }
+        
+        let layer = view.layer.presentation() ?? view.layer
+        let size = layer.bounds.size
+        let layerMiddle = CGPoint(x: size.width * layer.anchorPoint.x, y: size.height * layer.anchorPoint.y)
+        
+        var newTransform = transform.translatedBy(x: layer.position.x, y: layer.position.y)
+        newTransform = view.transform.concatenating(newTransform)
+        newTransform = newTransform.translatedBy(x: -layerMiddle.x, y: -layerMiddle.y)
         
         let ignore = shouldIgnore(view: view)
-        let redact = shouldRedact(view: view, redactText: redactText, redactImage: redactImage)
+        let redact = shouldRedact(view: view, redactOptions: redactOptions)
         
         if !ignore && redact {
-            redacting.append(RedactRegion(rect: rectInWindow, color: self.color(for: view)))
+            redacting.append(RedactRegion(size: size, transform: newTransform, type: .redact, color: self.color(for: view)))
             return
-        } else if hasBackground(view) {
-            if rectInWindow == area {
+        } else if isOpaque(view) {
+            let finalViewFrame = CGRect(origin: .zero, size: size).applying(newTransform)
+            if isAxisAligned(newTransform) && finalViewFrame == rootFrame {
+                //Because the current view is covering everything we found so far we can clear `redacting` list
                 redacting.removeAll()
```

Comment:
It seems that you got the view hierarchy wrong.

> It seems like we can early return from here instead of traversing the subviews

This is a view that is occupying the whole screen, we just need to ditch everything that come before it, but we definitely need to check its subviews.

> we're processing lots of view hierarchy that isn't visible by continuing on below from here to the loop through the subviews

Subviews are not below but above their parent view; we start from the backmost view in the view hierarchy and move forward. The only thing that would makes us stop traversing the view hierarchy is a view that is marked to be ignored.

Reply:
> It seems like we can early return from here instead of traversing the subviews

This is a view that is occupying the whole screen, we just need to ditch everything that come before it, but we definitely need to check its subviews.

> we're processing lots of view hierarchy that isn't visible by continuing on below from here to the loop through the subviews

Subviews are not below but above their parent view; we start from the backmost view in the view hierarchy and move forward. The only thing that would makes us stop traversing the view hierarchy is a view that is marked to be ignored.